From 4e79ddb569e4a2fb90aa8d155fe3a8451407fd62 Mon Sep 17 00:00:00 2001
From: Aleksei Nikiforov <darktemplar@basealt.ru>
Date: Mon, 3 Feb 2020 11:34:05 +0300
Subject: [PATCH 046/171] xcb: ensure that available glx version is greater
 than requested one or equal to it

Otherwise xcb plugin may report that it's supported
and later it'd be unable to create opengl context
when requested due to too low xcb protocol version available,
and thus some important xcb functions missing.

Without such change when Qt application is running under
x2go it exits with fatal error due to mismatch in
requested and supported glx protocol versions.
Qt requests glx protocol version 1.4, while x2go currently
reports that it support glx protocol but only version 1.2,
but Qt uses functions such as glXChooseFBConfig and
glXGetVisualFromFBConfig which are available only starting with
glx protocol version 1.3 according to glx documentation.

With this change, when Qt application is running under x2go,
xcb plugin reports that it's not supported and egl plugin
is used instead. And egl plugin successfully allows to work
with opengl, and thus Qt application is running normally.

Change-Id: I5c0fb10bd328da68054bfca8e8efde1144789566
Reviewed-by: Laszlo Agocs <laszlo.agocs@qt.io>
(cherry picked from commit b5feb28d63f763f1f830af1f7586712aa45ef4be)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 .../xcb/gl_integrations/xcb_glx/qxcbglxintegration.cpp    | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/plugins/platforms/xcb/gl_integrations/xcb_glx/qxcbglxintegration.cpp b/src/plugins/platforms/xcb/gl_integrations/xcb_glx/qxcbglxintegration.cpp
index 6814dbd844..fa708fbf26 100644
--- a/src/plugins/platforms/xcb/gl_integrations/xcb_glx/qxcbglxintegration.cpp
+++ b/src/plugins/platforms/xcb/gl_integrations/xcb_glx/qxcbglxintegration.cpp
@@ -48,6 +48,7 @@
 #include "qxcbscreen.h"
 #include "qglxintegration.h"
 
+#include <QtCore/QVersionNumber>
 #include <QtGui/QOpenGLContext>
 
 #include "qxcbglxnativeinterfacehandler.h"
@@ -59,6 +60,9 @@
 QT_BEGIN_NAMESPACE
 
 #if QT_CONFIG(xcb_glx)
+  #define QT_XCB_GLX_REQUIRED_MAJOR 1
+  #define QT_XCB_GLX_REQUIRED_MINOR 3
+
   #if XCB_GLX_MAJOR_VERSION == 1 && XCB_GLX_MINOR_VERSION < 4
     #define XCB_GLX_BUFFER_SWAP_COMPLETE 1
     typedef struct xcb_glx_buffer_swap_complete_event_t {
@@ -113,7 +117,9 @@ bool QXcbGlxIntegration::initialize(QXcbConnection *connection)
     auto xglx_query = Q_XCB_REPLY(xcb_glx_query_version, m_connection->xcb_connection(),
                                   XCB_GLX_MAJOR_VERSION,
                                   XCB_GLX_MINOR_VERSION);
-    if (!xglx_query) {
+    if ((!xglx_query)
+        || (QVersionNumber(xglx_query->major_version, xglx_query->minor_version)
+            < QVersionNumber(QT_XCB_GLX_REQUIRED_MAJOR, QT_XCB_GLX_REQUIRED_MINOR))) {
         qCWarning(lcQpaGl) << "QXcbConnection: Failed to initialize GLX";
         return false;
     }
-- 
2.31.1

